CONTRIBUTING.md

```diff
@@ -1,36 +1,202 @@
+
+### With Breaking Changes
+
+Sometimes, we don't merge pull requests with breaking changes immediately upon approval. Since a breaking change will cause Noir to bump to the next "minor" version, we might want to land some fixes in "patch" releases before we begin working on that next breaking version.
+
+## Merging
+
+Once approved by the required number of team members, the pull request can be merged into the `master` branch. Sometimes, especially for external contributions, the final approver may merge the pull request instead of the submitter.
+
+We use "squash merging" to merge all pull requests. This will cause all commits to be combined into one commit—another reason we want to keep pull requests small & focused.
+
+### Squash Merging
+
+When squash merging, we can keep intermediate Conventional Commits by adding them to the body of the commit message; however, the GitHub UI adds a `*` character before each commit message and our releaser bot won't parse that.
+
+When squashing, you want to update both the title of the commit to be a good Conventional Commit and adjust the body to contain any other Conventional Commits that you want to keep (not prefixed with `*`) and delete any extra information. We also keep any "Co-authored-by:" lines at the bottom of the commit if the change was done by multiple authors. If "Co-authored-by:" lines appear due to accepted PR suggestions, it's good to delete them so the author gets full credit for the change.
+
+Our overall approach to squashing is to be mindful of the impact of each commit. The commits populate our Changelog, so it's important to properly convey to Noir consumers what changes have happened. It is also a record that we and others will review in the future. Thus, we want to attribute the change to its correct authors and provide useful information that future contributors need.
+
+For example, given the default squash commit message:
+
+```
+feat(syntax): Implement first-class functions (#123)
+
+* clippy
+
+* formatting
+
+* chore(ci): Use correct rust version
+
+* review comment changes
+
+* Accepted suggestion from @phated
+
+* clippy
+```
+
+The person merging would remove extrenous messaging and keep only the relavent Conventional Commits:
+
+```
+feat(syntax): Implement String data type (#123)
+
+chore(ci): Use correct rust version
+```
+
+Additional Conventional Commits can be added before squashing if they improve the Changelog or commit history:
+
+```
+feat(syntax): Implement String data type (#123)
+
+chore(ci): Use correct rust version
+chore(clippy): Update all println calls to use single argument syntax
+```
+
+### Merge Checklist
+
+Before merging, you should mentally review these questions:
+
+- Is continuous integration passing?
+- Do you have the required amount of approvals?
+- Does anyone else need to be pinged for thoughts?
+- Will this cause problems for our release schedule? For example: maybe a patch release still needs to be published.
```

Comment:
While releases are automatically built from the `master` branch, we could merge our PRs into a new `dev` branch by default (that's periodically copied over to `master` depending on release cadence).

That way we can enjoy the flexibility to merge specific patching PRs directly into `master` when needed.

Reply:
Alternatively we can also do [merge queues](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/merging-a-pull-request-with-a-merge-queue), but that seems more manually involved than I'd like.

Reply:
Having a dev/master branch gives a lot of flexibility but it also comes at a cost.
I think the stable/nightly releases scheme avoid this cost, since the stable release is stable (because it was tested for a long time), there is not need to apply critical bug fixes to it (else it means the release was not stable), so you do not need the flexibility to cherry-pick a bug fix.

Reply:
I'd rather keep development on master and move releases to release branches if anything. Releases could be tagged commits on master to avoid clogging up branches, and we can still retroactively branch off these commits to add patches later if needed.

Reply:
Agreed on keeping development on master (at least for the near future) rather than having a dev branch.

Reply:
I agree with that `dev` branch *shouldn't* be necessary because master will be the rolling development trunk where nightlies will be produced and release-please creates the tags on the release commit for us (and we can backport changes in a branch if we need to, but that will require some additional CI changes).